object.go

```diff
@@ -4,12 +4,127 @@ import (
 	"encoding/xml"
 	"io"
 	"net/http"
+	"crypto/sha256"
 	"time"
+	"regexp"
+	"bufio"
+	"fmt"
+	"strconv"
+	"errors"
 
 	"github.com/gorilla/mux"
 	"github.com/sirupsen/logrus"
 )
 
+var (
+	// chunkValidator is a regexp for validating a chunk "header" in the
+	// request body of a multi-chunk upload
+	chunkValidator = regexp.MustCompile(`^([0-9a-fA-F]+);chunk-signature=([0-9a-fA-F]+)`)
+
+	// InvalidChunk is an error returned when reading a multi-chunk object
+	// upload that contains an invalid chunk header or body
+	InvalidChunk = errors.New("invalid chunk")
+)
+
+// Reads a multi-chunk upload body
+type chunkedReader struct {
+	body io.ReadCloser
+	lastChunk []byte
+	bufBody *bufio.Reader
+	
+	signingKey []byte
+	lastSignature string
+	timestamp string
+	date string
+	region string
+}
+
+func newChunkedReader(body io.ReadCloser, signingKey []byte, seedSignature, timestamp, date, region string) *chunkedReader {
+	return &chunkedReader {
+		body: body,
+		lastChunk: nil,
+		bufBody: bufio.NewReader(body),
+
+		lastSignature: seedSignature,
+		timestamp: timestamp,
+		date: date,
+		region: region,
+	}
+}
+
+func (c *chunkedReader) Read(p []byte) (n int, err error) {
+	if c.lastChunk == nil {
+		if err := c.readChunk(); err != nil {
+			return 0, err
+		}
+	}
+
+	n = copy(p, c.lastChunk)
+
+	if n == len(c.lastChunk) {
+		c.lastChunk = nil
+	} else {
+		c.lastChunk = c.lastChunk[n:]
+	}
+
+	return n, nil
+}
+
+func (c *chunkedReader) readChunk() error {
+	// step 1: read the chunk header
+	line, err := c.bufBody.ReadString('\n')
+	if err != nil {
+		if err == io.EOF {
+			return err
+		}
+		return InvalidChunk
+	}
+
+	match := chunkValidator.FindStringSubmatch(line)
+	if len(match) == 0 {
+		return InvalidChunk
+	}
+
+	chunkLengthHexStr := match[1]
+	chunkSignature := match[2]
+
+	chunkLength, err := strconv.ParseUint(chunkLengthHexStr, 16, 32)
+	if err != nil {
+		return InvalidChunk
+	}
+
+	// step 2: read the chunk body
+	chunk := make([]byte, chunkLength)
+	_, err = io.ReadFull(c.bufBody, chunk)
```

Comment:
minor suggestion: I don't know if this is actually any faster, but you could use `sha256.New()` to get a writer and pass it to io.TeeReader[1] to hash the data while you copy it:
```
chunk := make([]byte, chunkLength)
hash := sha256.New()
_, err = io.ReadFull(io.TeeReader(hash, c.bufBody), chunk)
stringToSign := fmt.Sprintf(..., hash.Sum(nil))
```

Reply:
This is a little tricky. I'd have to use `io.MultiWriter`, change `chunk` to something that implements `io.Writer`, and nix using `io.ReadFull`. Given the uncertain performance implications, I'll just keep the current setup if that's ok with you.